### DISCUSSION TITLE
불변성에 대한 설명과 어떤 장단점이 있는지, 그리고 불변객체를 만드는 방법에 대해 설명해주세요

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#17] [Clean Code]
글쎄요 멀까요?

**Comment by kunsanglee** [comment]
불변성(Immutability)은 데이터의 상태가 생성 후에 변경되지 않는 특성을 말합니다. 

### 장점
- 불변 객체는 여러 스레드에 의해 동시에 사용되어도 상태가 변경되지 않으므로, 동시성 문제 없이 안전하게 사용할 수 있습니다. 
- 객체의 상태가 변경되지 않기 때문에, 코드의 동작이 더 예측 가능해지고 이해하기 쉬워집니다. 
- 상태 변경으로 인한 부작용(side effects)이 없기 때문에, 버그 발생 가능성이 줄어듭니다. 
- 불변 객체는 자주 재사용될 수 있으므로, 메모리 사용 효율성이 향상될 수 있습니다. 예를 들어, 같은 값을 가지는 불변 객체는 하나의 인스턴스만 생성하고 여러 곳에서 공유할 수 있습니다.

### 단점
- 객체의 상태를 변경할 때마다 새로운 객체를 생성해야 하므로, 잦은 상태 변경이 있는 경우 메모리 사용량이 증가할 수 있습니다. 
- 새로운 객체를 생성하고 이전 객체를 가비지 컬렉션 대상으로 만드는 과정에서 성능 저하가 발생할 수 있습니다.
- 불변 객체를 사용할 때는 상태 변경을 위해 새로운 객체를 반환하는 방식을 사용해야 하므로, 구현이 복잡해질 수 있습니다.

**Comment by seunghye218** [comment]
불변성은 객체의 상태가 생성된 후 변경될 수 없는 특성을 말합니다.

장점으로 

첫 번째로 불변 객체는 스레드 안전성이 높습니다. 동시에 접근하더라도 상태가 변경되지 않기 때문입니다.

두 번째로 상태가 변경되지 않기 때문에 오류 발생 가능성이 낮습니다.

마지막으로 불변 객체는 여러 스레드에서 공유될 수 있어 메모리 사용이 효율적입니다. 

하지만  복잡한 도메인 모델을 불변 객체로 구현하기 어려울 수 있고 불변 객체의 상태를 변경해야 하는 경우 새로운 객체를 만들어야 하므로 코드가 복잡할 수 있습니다.

**Comment by GIVEN53** [comment]
생성 시점에 결정된 상태가 다른 상태로 변경되지 않는 특성

thread safe
동작을 예측하기 쉬움
재사용 -> 메모리 사용 효율

잦은 상태 변경 -> 메모리 사용량 증가
잦은 GC로 인한 성능 저하
구현이 복잡해질 수 있음
